src/main/java/org/apache/brooklyn/cloudfoundry/location/CloudFoundryPaasClient.java

```diff
@@ -18,66 +18,189 @@
  */
 package org.apache.brooklyn.cloudfoundry.location;
 
+import java.io.File;
 import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.List;
 import java.util.Map;
 
+import org.apache.brooklyn.cloudfoundry.entity.VanillaCloudfoundryApplication;
+import org.apache.brooklyn.cloudfoundry.utils.LocalResourcesDownloader;
+import org.apache.brooklyn.util.collections.MutableList;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.apache.brooklyn.util.exceptions.Exceptions;
+import org.apache.brooklyn.util.text.Strings;
 import org.cloudfoundry.client.lib.CloudCredentials;
 import org.cloudfoundry.client.lib.CloudFoundryClient;
+import org.cloudfoundry.client.lib.CloudFoundryException;
+import org.cloudfoundry.client.lib.StartingInfo;
+import org.cloudfoundry.client.lib.domain.CloudApplication;
+import org.cloudfoundry.client.lib.domain.CloudDomain;
+import org.cloudfoundry.client.lib.domain.Staging;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
 
 public class CloudFoundryPaasClient {
 
+    private static final Logger log = LoggerFactory.getLogger(CloudFoundryPaasClient.class);
+
+
     private final CloudFoundryPaasLocation location;
     private CloudFoundryClient client;
 
     public CloudFoundryPaasClient(CloudFoundryPaasLocation location) {
         this.location = location;
-        setUpClient();
     }
 
-    private void setUpClient() {
+    protected CloudFoundryClient getClient() {
         if (client == null) {
             CloudCredentials credentials =
                     new CloudCredentials(
                             location.getConfig(CloudFoundryPaasLocationConfig.ACCESS_IDENTITY),
                             location.getConfig(CloudFoundryPaasLocationConfig.ACCESS_CREDENTIAL));
-            client = new CloudFoundryClient(credentials, null);
+            client = new CloudFoundryClient(credentials,
+                    getTargetURL(location.getConfig(CloudFoundryPaasLocationConfig.CLOUD_ENDPOINT)),
+                    location.getConfig(CloudFoundryPaasLocationConfig.CF_ORG),
+                    location.getConfig(CloudFoundryPaasLocationConfig.CF_SPACE), true);
+
             client.login();
         }
+        return client;
     }
 
     public String deploy(Map<?, ?> params) {
-        String domain;
-        //TODO using client to deploy application
-        //client.createApplication(applicationName, buildpack, memory, disk,...)
-        //pushArtifact(applicationName, artifact);
+        ConfigBag appSetUp = ConfigBag.newInstance(params);
+        String artifactLocalPath =
+                getLocalPath(appSetUp.get(VanillaCloudfoundryApplication.ARTIFACT_PATH));
```

Comment:
It would not make sense. The `ConfigKey` which is used here, `ARTIFACT_PATH`, does not been read from `VanillaCloudfoundryApplication` because `CloudFoundryPaasClient` and `CloudFoundryPaasLocation` should be independent of any entity